### PR TITLE
Copp tmp br

### PR DESCRIPTION
**Pull request by noaOrMlnx**
**What I did**

**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: cfgmgr/coppmgr.cpp [hunk 1]
```diff
@@ -96,13 +96,38 @@ void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
     string trap_group = m_coppTrapConfMap[feature].trap_group;
     bool prev_group_state = checkTrapGroupPending(trap_group);
 
+    // update features cache
+    auto state = "disabled";
+    if (enable)
+    {
+        state = "enabled";
+    }
+    auto vect = m_featuresCfgTable[feature];
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp]: This statement contradicts with the find below. You will try to access the feature table with key even if it is not found leading to key being inserted. Please revisit the logic

File: cfgmgr/coppmgr.cpp [hunk 2]
```diff
@@ -96,13 +96,38 @@ void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
...
+        }
+        m_featuresCfgTable.at(feature) = vect;
+    }
+
     if (!enable)
     {
-        m_coppDisabledTraps.insert(feature);
+        if (m_coppTrapConfMap[feature].is_always_enabled != "true")
+        {
+            m_coppDisabledTraps.insert(feature);
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp]: Can we remove this m_coppDisabledTraps as it is of no purpose now? 
The variable is not used anywhere. Previously it was used to find if a trap is disabled. However, currently the logic is rewritten in such a way that we use always enabled or check if feature table has enable.

File: cfgmgr/coppmgr.cpp [hunk 3]
```diff
@@ -140,25 +165,61 @@ void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
...
 {
-    for (auto &m: m_coppDisabledTraps)
+    if (m_featuresCfgTable.find(feature) != m_featuresCfgTable.end())
     {
-        if (m_coppTrapConfMap.find(m) == m_coppTrapConfMap.end())
+        std::vector<FieldValueTuple> feature_fvs = m_featuresCfgTable[feature];
+        for (auto i: feature_fvs)
         {
-            continue;
+            if (fvField(i) == "state" && fvValue(i) == "enabled")
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp:161]: Feature table too has "always_enabled" field which also means enabled. Please take that into consideration.

File: cfgmgr/coppmgr.cpp [hunk 4]
```diff
@@ -264,12 +326,35 @@ CoppMgr::CoppMgr(DBConnector *cfgDb, DBConnector *appDb, DBConnector *stateDb, c
         }
     }
 
+    /* If there is a trap that has always_enabled = true, remove it from the disabledTraps list */
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp]: This entire code can be removed based on my above comment.

File: cfgmgr/coppmgr.cpp [hunk 5]
```diff
@@ -450,68 +587,168 @@ void CoppMgr::doCoppTrapTask(Consumer &consumer)
...
-                    FieldValueTuple fv(COPP_TRAP_ID_LIST_FIELD, trap_ids);
-                    fvs.push_back(fv);
-                    if (!checkTrapGroupPending(m_coppTrapConfMap[key].trap_group))
-                    {
-                        m_appCoppTable.set(m_coppTrapConfMap[key].trap_group, fvs);
-                        setCoppGroupStateOk(m_coppTrapConfMap[key].trap_group);
-                    }
-                    m_coppTrapConfMap.erase(key);
+
+                    m_coppDeletedTrapConfMap[key] = m_coppTrapConfMap[key];
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp]: I still do not understand the reason for null_cfg.
If the null_cfg is present it has just one key value pair which is null, null. We don't support hdel. So If a user sets the config again the user has to set the complete configuration. 
I don't think we need this data structure.

**Review by dgsudharsan — CHANGES_REQUESTED** [review]

File: cfgmgr/coppmgr.cpp (hunk 1 above)
Comment (noaOrMlnx) [reply cfgmgr/coppmgr.cpp]: Done

File: cfgmgr/coppmgr.cpp (hunk 3 above)
Comment (noaOrMlnx) [reply cfgmgr/coppmgr.cpp:161]: Done

File: cfgmgr/coppmgr.cpp (hunk 5 above)
Comment (noaOrMlnx) [reply cfgmgr/coppmgr.cpp]: Done

File: cfgmgr/coppmgr.cpp [hunk 6]
```diff
@@ -140,25 +151,50 @@ void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
...
-        if (m_coppTrapConfMap.find(m) == m_coppTrapConfMap.end())
+        std::vector<FieldValueTuple> feature_fvs = m_featuresCfgTable[feature];
+        for (auto i: feature_fvs)
         {
-            continue;
+            if (fvField(i) == "state" && fvValue(i) == "enabled")
+            {
+                return true;
+            }
+            if (fvField(i) == "state" && fvValue(i) == "always_enabled")
```
Comment (dgsudharsan) [inline cfgmgr/coppmgr.cpp:165]: Can we combine both the if statements?
&& (fvValue(i) == "enabled" || fvValue(i) == "always_enabled")

**Comment by dgsudharsan** [comment]
One more minor comment. Please remove disabled trap from header.

https://github.com/noaOrMlnx/sonic-swss/blob/0feeafa400bc17b571f6c5a353e423fc9868ddc6/cfgmgr/coppmgr.h#L75

**Pull request closed without merging**